axlearn/common/optimizers.py

```diff
@@ -1033,6 +1033,113 @@ def update_fn(updates, state, params=None):
     )
 
 
+class SkipClipState(NamedTuple):
+    """State returned by functions in skip_and_clip_by_global_norm()."""
+
+    nonvalid_count: Union[Tensor, TensorSpec]  # Number of non-valid steps.
+    inner_state: Any  # State of the inner PartitionedGradientTransformation.
+
+
+def skip_and_clip_by_global_norm(
+    inner: ConfigOr[PartitionedGradientTransformation],
+    *,
+    drop_norm: Optional[float] = None,
+    max_norm: Optional[float] = None,
+    eps: float = 1e-8,
+) -> PartitionedGradientTransformation:
+    """Skip updates when global norm >= drop_norm, otherwise clip the global norm.
+    If we detect abnormal gradients that have global norm >= drop_norm, we skip the gradient updates
+    and state updates. Otherwise we scale the gradients s.t. global norm <= max_norm, and apply the
+    wrapped gradient transformation `inner`. Note the difference compared to clip_by_global_norm()
+    is that this version skips all updates while clip_by_global_norm() still performs parameter
+    updates and optimizer state updates.
+    Example usage:
```

Comment:
Can we fix the docstring spacing (newline before example, args, returns)?

Reply:
Done!